File: src/media115/scraper/registry.py
"""Provider registry with priority-based fallback."""
from __future__ import annotations

from pathlib import Path

from media115.scraper.provider import BaseProvider


# Registry: media_type → list of providers sorted by priority
_PROVIDERS: dict[str, list[BaseProvider]] = {}


def register(provider: BaseProvider):
    """Register a provider for its supported types."""
    for media_type in provider.supported_types:
        if media_type not in _PROVIDERS:
            _PROVIDERS[media_type] = []
        _PROVIDERS[media_type].append(provider)
        _PROVIDERS[media_type].sort(key=lambda p: p.priority)


def get_providers(media_type: str) -> list[BaseProvider]:
    """Get registered providers for a media type, sorted by priority."""
    return _PROVIDERS.get(media_type, [])


def scrape(media_type: str, query: str, filename: str, out_dir: Path, **kwargs) -> dict:
    """Try providers in priority order. First success wins."""
    from media115.log import get_logger
    logger = get_logger()

    providers = get_providers(media_type)
    if not providers:
        return {"status": "error", "error": f"No providers for type: {media_type}"}

    for provider in providers:
        try:
            logger.debug("SCRAPE     trying %s for %s '%s'", provider.name, media_type, query[:40])
            result = provider.scrape(query, filename, out_dir, **kwargs)
            if result and result.get("status") == "ok":
                result["_source"] = provider.name
                return result
            logger.debug("SCRAPE     %s returned %s", provider.name, result.get("status", "?"))
        except Exception as e:
            logger.warning("SCRAPE     %s failed: %s", provider.name, e)
            continue

    return {"status": "not_found", "query": query}

File: internal/scraper/registry.go
package scraper

import (
	"log/slog"
	"sort"
	"sync"
)

var (
	mu        sync.RWMutex
	providers = map[string][]Provider{} // mediaType → sorted by priority
)

// Register adds p to the registry for every media type it declares.
// Providers with a lower Priority() value run first.
func Register(p Provider) {
	mu.Lock()
	defer mu.Unlock()
	for _, t := range p.SupportedTypes() {
		providers[t] = append(providers[t], p)
		sort.Slice(providers[t], func(i, j int) bool {
			return providers[t][i].Priority() < providers[t][j].Priority()
		})
	}
}

// Reset clears all registered providers. Intended for use in tests.
func Reset() {
	mu.Lock()
	defer mu.Unlock()
	providers = make(map[string][]Provider)
}

// GetProviders returns the registered providers for the given media type,
// ordered by priority (lowest first).  Returns nil when none are registered.
func GetProviders(mediaType string) []Provider {
	mu.RLock()
	defer mu.RUnlock()
	return providers[mediaType]
}

// Scrape tries providers for mediaType in priority order and returns the first
// successful result.  If no provider succeeds it returns Status "not_found".
func Scrape(mediaType, query, filename, outDir string, opts ScrapeOpts, logger *slog.Logger) *ScrapeResult {
	if logger == nil {
		logger = slog.Default()
	}

	prov := GetProviders(mediaType)
	if len(prov) == 0 {
		return &ScrapeResult{Status: "error", Error: "no providers for type: " + mediaType}
	}

	for _, p := range prov {
		logger.Debug("SCRAPE trying", "provider", p.Name(), "type", mediaType, "query", query)
		result, err := p.Scrape(query, filename, outDir, opts)
		if err != nil {
			logger.Warn("SCRAPE failed", "provider", p.Name(), "error", err)
			continue
		}
		if result != nil && result.Status == "ok" {
			result.Source = p.Name()
			return result
		}
		if result != nil {
			logger.Debug("SCRAPE not matched", "provider", p.Name(), "status", result.Status)
		}
	}

	return &ScrapeResult{Status: "not_found"}
}
